### PR TITLE
Implement RSA encryption (with update for current rewrite structure)

### PR DESCRIPTION
**Pull request by en1gma713**
The same as my last PR but it works this time and is up-to-date on the file structure

### CODE REVIEW

File: src/lib/net/encryption/src/key.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+use rsa::{RsaPublicKey, RsaPrivateKey, pkcs1::EncodeRsaPublicKey, Pkcs1v15Encrypt};
+use rand::rngs::OsRng;
+use crate::errors;
+
+pub struct KeyPair {
+    pub public_key: RsaPublicKey, //public key side of the keypair (for encryption)
+    pub encoded_public_key: der::Document, //the public key encoded in ASN.1 DER format for publishing to the client
+    pub private_key: RsaPrivateKey, //private key side of the keypair (for decryption)
```
Comment (Asurar0) [inline src/lib/net/encryption/src/key.rs:8]: When documenting field the actual syntax should be:
```rust
/// public key side of the keypair (for encryption)
pub public_key: RsaPublicKey
```
As it ties the documentation comment to the field that is then viewable with Rust-analyzer/LSP

File: src/lib/net/encryption/Cargo.toml [hunk 2]
```diff
@@ -5,3 +5,6 @@ edition = "2021"
 
 [dependencies]
 thiserror = { workspace = true }
+rsa = "0.9.6"
+der = "0.7.9"
+rand = "0.8.5"
```
Comment (Asurar0) [inline src/lib/net/encryption/Cargo.toml:10]: These dependencies should be placed in the workspace (project root) `Cargo.toml` and then imported as 
`crate = { workspace = true }`

See https://github.com/ferrumc-rs/ferrumc/blob/rewrite/v3/src/lib/README.MD

**Review by Asurar0 — CHANGES_REQUESTED** [review]

**Comment by en1gma713** [comment]
fixed it

**Comment by en1gma713** [comment]
facing version conflicts between rand and rand_core now

**Comment by ReCore-sys** [comment]
@en1gma713 we've added the networking stack now, if you'd be able to sort out the conflicts and make sure all of your stuff works ok with the new code, im happy to merge

**Comment by ReCore-sys** [comment]
No work is being done one this and the codebase has evolved to the point where this PR is unlikely to be easier to fix than just writing from scratch again

**Pull request closed without merging**